File: python/bob/bob.py
from collections import OrderedDict


def isSilence(what):
    return what.strip() == ''


def isShouting(what):
    return what.isupper()


def isQuestion(what):
    return what.strip().endswith('?')


def isAnything(_):
    return True


lookup = OrderedDict([(isSilence, 'Fine. Be that way!'), (
    isShouting, 'Whoa, chill out!'), (isQuestion, 'Sure.')])


def hey(what):
    return next(
        (result for condition, result in lookup.items()
         if condition(what)), 'Whatever.')

File: java/bob/src/main/java/Bob.java
import java.util.Optional;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Stream;

public class Bob {
    public String hey(String input) {
        final String cleanedInput = (input != null ? input : "");

        return Stream.of(silence, shouting, question)
                .map(listener -> listener.apply(cleanedInput))
                .filter(response -> response.isPresent())
                .map(response -> response.get())
                .findFirst()
                .orElse("Whatever.");
    }

    private static Function<String, Optional<String>> when(Predicate<String> matcher, String response) {
        return then.apply(response).apply(matcher);
    }

    private static Function<String, Function<Predicate<String>, Function<String, Optional<String>>>> then
            = response -> matcher -> phrase -> Optional.of(phrase).filter(matcher).map(ignore -> response);

    private static Predicate<String> isSilent = input -> input.trim().isEmpty();

    private static Predicate<String> isShouting = input -> input.toUpperCase().equals(input)
            && !input.equals(input.toLowerCase());

    private static Predicate<String> isQuestion = input -> input.endsWith("?");

    private static Function<String, Optional<String>> silence = when(isSilent, "Fine. Be that way!");
    private static Function<String, Optional<String>> shouting = when(isShouting, "Whoa, chill out!");
    private static Function<String, Optional<String>> question = when(isQuestion, "Sure.");
}